Added hurricane test case with local time-stepping

gcapodag opened the pull request:
Created a test case to simulate the storm surge of hurricane Sandy using a local time-stepping method instead of Runge-Kutta 4. This new test case is called `hurricane_lts` and it is based on the existing `hurricane` test case.

As part of this PR, I have added `lts_region.py` in `compass/ocean/mesh` which takes care of adding LTS regions to a culled mesh. The differences with `hurricane` are:
1. After the cull mesh step another step is called `lts_regions` that invokes the aforementioned script and does two things. First, it computes the array `LTSRegion` and injects into `culled_mesh.nc`. Then it modifies the `culled_graph.info` to revise the load balancing according to the LTS regions computed.
2. In the init step, there is an additional step that computes the topographic drag. 
3. I removed the namelist [here](https://github.com/gcapodag/compass/blob/72ca95ca863d3b4a25094a86afdc9efa5c36d0d9/compass/ocean/tests/hurricane_lts/forward/forward.py#L48-L49) and included that information into the forward namelist, which is different compared to the one in the `hurricane` test case, and so is the `streams.ocean`.

Note that currently, doing `compass run` in the sandy directory produces [this](https://github.com/E3SM-Project/E3SM/issues/5510) error, and this happens with both `hurricane` and `hurricane_lts`.

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist

Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [x] The `E3SM-Project` submodule has been updated with relevant E3SM changes
* [ ] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes
* [ ] New tests have been added to a test suite

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

xylar commented:
Thanks @gcapodag!  This will be an amazing new capability.

Could you provide build instructions for MPAS-Ocean required to test this test group?

xylar commented:
I have spoken with both @gcapodag and @sbrus89 about this work.  I do feel strongly that `hurricane_lts` shouldn't be its own test group and instead should be part of `hurricane`.  I am happy to help with that process but am hoping you 2 can get the process started since I'm away this week at EGU.

I encouraged @gcapodag to make this PR so we could start the discussion here.

xylar commented:
@gcapodag, I added the full checklist back into the PR description.  You will eventually need to add documentation.  I don't think you have added `hurricane_lts` tests to any test suites, though you had checked that box.  You will need tell us here what testing you have run. And presumably a change in E3SM is still required to support LTS, or has that already been merged?

xylar commented:
One more thing, you will need to rebase and bring in we linting that was added earlier this year.  A rebase and then some linting will be needed, but that probably makes sense to do with the merge of `hurricane` and `hurricane_lts`.

I'm happy to help with any trouble you run into regarding linting.

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,362 @@
+                   graph_info='culled_graph.info', num_interface=2)
+
+
+def get_fine_region(region):
```
It would be better if these regions were defined in one or more data files rather than hard-coded.  geojson seems like the most natural format but NetCDF would be fine.  The file would be small enough that it could go in the code but I would still put it in the database instead.

gcapodag replied:
Done with latest commit.

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,362 @@
+def label_mesh(fine_region, mesh, graph_info, num_interface):   # noqa: C901
+
+    # read in mesh data
+    ds = xarray.open_dataset(mesh)
+    nCells = ds['nCells'].size
+    nEdges = ds['nEdges'].size
+    areaCell = ds['areaCell'].values
+    cellsOnEdge = ds['cellsOnEdge'].values
+    edgesOnCell = ds['edgesOnCell'].values
+    latCell = ds['latCell']
+    lonCell = ds['lonCell']
+
+    # start by setting all cells to coarse
+    LTSRgn = [2] * nCells
+
```
I'd like to keep all variables lowercase with underscores, even the mesh ones.

gcapodag replied:
Done with latest commit.

xylar reviewed (commented):
A few minor comments on `LTSRegionsStep`.

gcapodag commented:
> Thanks @gcapodag! This will be an amazing new capability.
> 
> Could you provide build instructions for MPAS-Ocean required to test this test group?

Hi Xylar, yes, the build instructions are: `make gnu-cray USE_LTS=true DEBUG=false OPENMP=false`

gcapodag commented:
> @gcapodag, I added the full checklist back into the PR description. You will eventually need to add documentation. I don't think you have added `hurricane_lts` tests to any test suites, though you had checked that box. You will need tell us here what testing you have run. And presumably a change in E3SM is still required to support LTS, or has that already been merged?

Hi Xylar, thanks for adding the list back in. The changes in E3SM to support LTS have been approved and are ready to be merged, see [this](https://github.com/E3SM-Project/E3SM/pull/5460) PR, though have not yet been merged. What is to be merged is in [this](https://github.com/gcapodag/E3SM/tree/gcapodag/mpaso/lts) branch.

gcapodag commented:
> One more thing, you will need to rebase and bring in we linting that was added earlier this year. A rebase and then some linting will be needed, but that probably makes sense to do with the merge of `hurricane` and `hurricane_lts`.
> 
> I'm happy to help with any trouble you run into regarding linting.
> 
> **Update** it seems like at least some linting has been done already, which is great! I don't understand why CI isn't running, though, so I still think a rebase would be good.

Yes, I made sure to comply with all the flake8 requirements, although I believe `hurricane` does not at the moment, so if a merging is required we should take care of that too.

gcapodag commented:
I modified the `hurricane` test group so that now as part of that group there are three new test case:`mesh_lts`, `init_lts` and `sandy_lts`, listed after the non-LTS ones, as @xylar suggested. I also removed the `hurricane_lts` test group.

xylar commented:
I tried testing this out a bit but wasn't able to run `compass list`.  It seems like maybe `compass.ocean.tests.hurricane.lts` is missing.  Did it not get committed?

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+        """
+        super().setup()
+
+        config = self.cull_mesh_step.config
+        self.cpus_per_task = config.getint('spherical_mesh',
+                                           'cull_mesh_cpus_per_task')
+        self.min_cpus_per_task = config.getint('spherical_mesh',
+                                               'cull_mesh_min_cpus_per_task')
```
It doesn't look like you're doing anything in parallel here so you don't need to set `cpus_per_task this way.  You can let things default to 1 task and 1 cpu per task.
```suggestion
        config = self.config
```
There is also no reason to use `config` from the cull_mesh step instead of from this step, since they will have the same contents (all steps in a test case have the same config options).

gcapodag replied:
Thanks, done.

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+        subdir : str, optional
+            the subdirectory for the step.  The default is ``name``
+        """
+        super().__init__(test_case, name=name, subdir=subdir,
+                         cpus_per_task=None, min_cpus_per_task=None)
```
For this step, it makes sense to default to 1 task and 1 cpu per task.
```suggestion
        super().__init__(test_case, name=name, subdir=subdir)
```

gcapodag replied:
Done.

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+
+    Attributes
+    ----------
+    cull_mesh_step : CullMeshStep
+        The culled mesh step containing input files to this step
```
This needs to be the full path so it shows up right in the auto-generated documentation.
```suggestion
    cull_mesh_step : compass.ocean.mesh.cull.CullMeshStep
        The culled mesh step containing input files to this step
```

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+
+        self.lts_fine_regions_file = 'hurricane_lts_fine_regions.nc'
+
+        self.add_input_file(
+            filename=self.lts_fine_regions_file,
+            target='hurricane_lts_fine_regions.nc',
+            database='hurricane')
```
You will also need to add `culled_mesh.nc` as an input:
``` python
        cull_path = self.cull_mesh_step.path
        target = os.path.join(cull_path, 'culled_mesh.nc')
        self.add_input_file(filename='culled_mesh.nc', work_dir_target=target)
```

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+        self.min_cpus_per_task = config.getint('spherical_mesh',
+                                               'cull_mesh_min_cpus_per_task')
+
+        self.lts_fine_regions_file = 'hurricane_lts_fine_regions.nc'
```
This should be defined in `__init__()` instead of here and added to the attributes of the class.  Then, you should use it throughout the class rather than the hard-coded name.  (Or just stick with the hard-coded name, but please don't do a mix of both.)

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+from compass.step import Step
+
+
+class LTSRegionsStep(Step):
+    """
+    A step for adding LTS regions to a global MPAS-Ocean mesh
```
This class has several files that seem very specific to the `Hurricane` test group so it's not clear to me that it belongs in the general ocean framework.

gcapodag replied:
I think we resolved this during our call.

xylar reviewed (changes requested):
I just got through `lts_regions.py` so far, but there's a lot that we can discuss.

xylar commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+        file_name = 'hurricane_lts_fine_regions.nc'
+        fine_rgn = nc.Dataset(file_name, 'r', format='NETCDF4_64BIT')
+
+        region = 'east_us_coast'
```
Presumably, this should be an input (e.g. a config option) instead of being hard-coded?

gcapodag replied:
Eventually we will have an automated framework to set the LTS regions so we would not need to read them from an input file anymore. For the time being, if that's OK, we should probably keep it specific to hurricane.

xylar commented on compass/ocean/tests/hurricane/forward/forward.py:
```diff
@@ -38,15 +45,26 @@ def __init__(self, test_case, mesh, init, name='forward', subdir=None):
-            'compass.ocean.tests.hurricane.forward', 'namelist.ocean')
-        self.add_streams_file(
-            'compass.ocean.tests.hurricane.forward', 'streams.ocean')
+        if use_lts is False:
```
This is the preferred way:
```suggestion
        if not use_lts:
```
Though I nearly always do the `if <blah>: ... else:` version rather than the `if not <blah>: ... else:` because in this latter case the `else` has a weird double-negative meaning.

gcapodag replied:
I changed it to `if use_lts:` `else:`

xylar commented on compass/ocean/tests/hurricane/forward/forward.py:
```diff
@@ -55,6 +73,13 @@ def __init__(self, test_case, mesh, init, name='forward', subdir=None):
             filename='atmospheric_forcing.nc',
             work_dir_target=f'{init.path}/interpolate/atmospheric_forcing.nc')
+
+        if use_lts is True:
```
```suggestion
        if use_lts:
```

gcapodag replied:
Done.

xylar commented on compass/ocean/tests/hurricane/init/__init__.py:
```diff
@@ -32,17 +37,29 @@ def __init__(self, test_group, mesh, storm):
+
+        self.mesh = mesh
+        self.use_lts = use_lts
+        if use_lts is True:
```
```suggestion
        if use_lts:
```

xylar commented on compass/ocean/tests/hurricane/init/__init__.py:
```diff
@@ -32,17 +37,29 @@ def __init__(self, test_group, mesh, storm):
         self.add_step(InterpolateAtmForcing(test_case=self, mesh=mesh,
                                             storm=storm))
+
+        if use_lts is True:
```
```suggestion
        if use_lts:
```

gcapodag replied:
Done.

gcapodag replied:
Done.

xylar commented on compass/ocean/tests/hurricane/mesh/__init__.py:
```diff
@@ -20,9 +22,16 @@ def __init__(self, test_group, mesh_name):
         self.mesh_name = mesh_name
-        name = 'mesh'
+        self.use_lts = use_lts
+        if use_lts is True:
```
```suggestion
        if use_lts:
```

gcapodag replied:
Done.

xylar commented on compass/ocean/tests/hurricane/mesh/__init__.py:
```diff
@@ -43,10 +52,17 @@ def __init__(self, test_group, mesh_name):
+
+        self.add_step(cull_mesh_step)
+
+        if use_lts is True:
```
```suggestion
        if use_lts:
```

gcapodag replied:
Done.

xylar commented on compass/ocean/tests/hurricane/forward/__init__.py:
```diff
@@ -34,15 +35,24 @@ def __init__(self, test_group, mesh, storm, init):
         """
         mesh_name = mesh.mesh_name
+        self.use_lts = use_lts
+        if use_lts is True:
```
```suggestion
        if use_lts:
```

xylar commented on compass/ocean/tests/hurricane/mesh/dequ120at30cr10rr2/__init__.py:
```diff
@@ -11,15 +11,12 @@ def build_cell_width_lat_lon(self):
-
         Returns
         -------
         cellWidth : numpy.array
             m x n array of cell width in km
-
         lon : numpy.array
             longitude in degrees (length n and between -180 and 180)
-
         lat : numpy.array
             longitude in degrees (length m and between -90 and 90)
```
I would leave the blank lines here as they were.  That's our formatting convention for docstrings.

gcapodag replied:
Done.

xylar reviewed (changes requested):
A few more small changes.  The rest looks excellent!  Very nice work.

gcapodag commented on compass/ocean/mesh/lts_regions.py:
```diff
@@ -0,0 +1,310 @@
+        test_case : compass.ocean.tests.global_ocean.mesh.Mesh
+            The test case this step belongs to
+
+        cull_mesh_step : cull_mesh_step
```
Done.

gcapodag replied:
Done.

gcapodag commented on compass/ocean/tests/hurricane/forward/__init__.py:
```diff
@@ -34,15 +35,24 @@ def __init__(self, test_group, mesh, storm, init):
+            Whether local time-stepping is to be used
         """
         mesh_name = mesh.mesh_name
+        self.use_lts = use_lts
```
Done.

xylar commented:
@gcapodag, please also move `lts_regions.py` to file into `hurricane` somewhere for now as well.  We can move it back to the ocean framework if it gets generalized in the future.

xylar commented on compass/ocean/tests/hurricane/forward/__init__.py:
```diff
@@ -34,15 +35,24 @@ def __init__(self, test_group, mesh, storm, init):
+            Whether local time-stepping is to be used
         """
         mesh_name = mesh.mesh_name
+
+        if use_lts:
+            storm = storm + '_lts'
         subdir = os.path.join(mesh_name, storm)
         super().__init__(test_group=test_group,
                          subdir=subdir,
                          name=storm)
```
These are both stylistic changes but I think it makes more sense to append `_lts` to the step name than to the `storm` variable.  I also have a strong preference for f-strings over string addition.

```suggestion

        if use_lts:
            name = f`{storm}_lts'
        else:
            name = storm
        subdir = os.path.join(mesh_name, name)
        super().__init__(test_group=test_group,
                         subdir=subdir,
                         name=name)
```

xylar replied:
As a results of this change, you will pass `storm = 'sandy'` to Analysis, but I think that will be cleaner anyway because you can revert several changes in that step with this change.  I don't think it's needed but you might have to pass `use_lts` to that step.

xylar commented on compass/ocean/tests/hurricane/analysis/__init__.py:
```diff
@@ -62,25 +64,25 @@ def setup(self):
         package = self.__module__
 
-        if self.storm == 'sandy':
+        if self.storm == 'sandy' or self.storm == 'sandy_lts':
```
The changes in this `if` would be reverted if you don't change `sandy` to `sandy_lts`.

gcapodag replied:
Hi Xylar, thanks for your feedback. If I leave `storm=sandy` regardless of whether lts is used, and pass that to Analysis, then wouldn't Analysis always pick up the non lts forward run? Now there is a `sandy_lts` test case folder where the forward output is saved when lts is on.

xylar reviewed (changes requested):
This looks excellent! I'll make sure I can set everything up but here are a few clean-up suggestions.

xylar reviewed (approved):
I'm happy!  I'm not able to set this up right now but only because the LCRC server is down.  I'll be happy to try again after things come back on Friday or Saturday but I'm happy to let things go through if @sbrus89 has time to review and test before that.

xylar commented:
Updates to the documentation are also still needed.

gcapodag commented:
> Ah, I spoke too soon. We need [E3SM-Project/E3SM#5460](https://github.com/E3SM-Project/E3SM/pull/5460) before this can go in so, it probably won't go in quite that quickly.

Yes, Mark needs to get back to Jon on a question he asked, but Mark is out this week so probably that will happen next week.

gcapodag commented:
> Updates to the documentation are also still needed.

I'd be happy to take care of that, are there any guidelines on best practices?

xylar commented:
> are there any guidelines on best practices?

https://mpas-dev.github.io/compass/latest/developers_guide/docs.html

You probably just need to edit the existing Hurricane documentation in both the user's and developer's guides (including the `ocean/api.rst` file in the latter) to make reference to LTS.  In the user's guide, make sure to mention that MPAS-Ocean needs to be compiled with `USE_LTS=true` in order to run the LTS tests.

gcapodag commented:
Hi @xylar and @sbrus89 I have amended the existing documentation for the hurricane test case including the new stuff related to LTS. I checked the three boxes above concerning the documentation. When you get a chance, please let me know if you would like me to edit/add anything to it. Thanks!

xylar reviewed (changes requested):
@gcapodag, I tried running this on Chrysalis with Intel and OpenMPI (and `USE_LTS=true`) but my forward run got stuck at `2012-10-14_16:39:36` and doesn't seem to advance.  I got PIO errors but they didn't seem to kill the run.

Unrelated to that, I have several changes to the docs and a few changes related to logging that I would appreciate if you would consider making.

gcapodag commented:
> @gcapodag, I tried running this on Chrysalis with Intel and OpenMPI (and `USE_LTS=true`) but my forward run got stuck at `2012-10-14_16:39:36` and doesn't seem to advance. I got PIO errors but they didn't seem to kill the run.
> 
> See: `/lcrc/group/e3sm/ac.xylar/compass_1.2/chrysalis/test_20230513/hurricane_lts`.
> 
> Unrelated to that, I have several changes to the docs and a few changes related to logging that I would appreciate if you would consider making.



> > Thanks @gcapodag! This will be an amazing new capability.
> > Could you provide build instructions for MPAS-Ocean required to test this test group?
> 
> Hi Xylar, yes, the build instructions are: `make gnu-cray USE_LTS=true DEBUG=false OPENMP=false` on Cori. That's where I did the testing.

Hi Xylar, I have always run it with `OPENMP=false`, see above instructions from Cori. Would you mind trying again? Thanks.

gcapodag commented on docs/developers_guide/ocean/api.rst:
```diff
@@ -337,6 +337,16 @@ test cases and steps
+   lts.mesh.lts_regions.LTSRegionsStep
+   lts.mesh.lts_regions.LTSRegionsStep.setup
+   lts.mesh.lts_regions.LTSRegionsStep.run
+   lts.mesh.lts_regions.LTSRegionsStep.label_mesh
```
Done.

gcapodag commented on docs/users_guide/ocean/test_groups/hurricane.rst:
```diff
@@ -114,6 +131,20 @@ This set reads in the observation station locations and finds the cells
+
+compute topographic wave drag step
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The reciprocal of the e-folding time, ``r_inv``, from the HyCOM model,
+is computed in this step. See this `reference`_ for details on the 
+computation. This coefficient is needed to account for the 
+topographic wave drag tendency in the model.
+
+.. _reference: https://doi.org/10.1175/JPO-D-15-0074.1
```
Done.

gcapodag commented on docs/users_guide/ocean/test_groups/hurricane.rst:
```diff
@@ -114,6 +131,20 @@ This set reads in the observation station locations and finds the cells
 
+.. _hurricane_init_lts:
+
+If the LTS option is selected for the init test case, the following
+additional step is carried out:
+
+compute topographic wave drag step
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The reciprocal of the e-folding time, ``r_inv``, from the HyCOM model,
```
Done.

gcapodag commented on docs/developers_guide/ocean/test_groups/hurricane.rst:
```diff
@@ -17,6 +17,16 @@ and provides the :py:func:`compass.ocean.tests.hurricane.mesh.dequ120at30cr10rr2
 method to specify the mesh resolution function using the :py:func:`mpas_tools.ocean.coastal_tools.coastal_refined_mesh()`
 function.
 
+lts_regions
+-----------
```
Done.

gcapodag commented on docs/developers_guide/ocean/test_groups/hurricane.rst:
```diff
@@ -17,6 +17,16 @@ and provides the :py:func:`compass.ocean.tests.hurricane.mesh.dequ120at30cr10rr2
 
+lts_regions
+-----------
+The class :py:class:`compass.ocean.tests.hurricane.lts.mesh.LTSRegionsStep` creates a 
```
Done.

gcapodag commented on compass/ocean/tests/hurricane/lts/mesh/lts_regions.py:
```diff
@@ -0,0 +1,321 @@
+                   graph_info='culled_graph.info', num_interface=2)
+
+
+def label_mesh(fine_region, mesh, graph_info, num_interface):   # noqa: C901
```
Done.

gcapodag commented on compass/ocean/tests/hurricane/lts/mesh/lts_regions.py:
```diff
@@ -0,0 +1,321 @@
+    lts_rgn = [2] * n_cells
+
+    # check each cell, if in the fine region, label as fine
+    print('Labeling fine cells...')
```
Done.

gcapodag commented on compass/ocean/tests/hurricane/lts/mesh/lts_regions.py:
```diff
@@ -0,0 +1,321 @@
+            delaware_coast = fine_rgn.variables['delaware_coast_pts'][:, :]
+            fine_region = Polygon(delaware_coast)
+
+        label_mesh(fine_region, mesh='culled_mesh.nc',
+                   graph_info='culled_graph.info', num_interface=2)
```
Done.

gcapodag commented on compass/ocean/tests/hurricane/lts/mesh/lts_regions.py:
```diff
@@ -0,0 +1,321 @@
+
+    mshnc.close()
+
+    sh_Command = 'paraview_vtk_field_extractor.py --ignore_time \
+                  -d maxEdges=0 -v allOnCells -f ' + ltsmsh_name + ' \
+                  -o lts_mesh_vtk'
+    sp.call(sh_Command.split())
```
Done. Thanks!

xylar commented:
@gcapodag, can you add build instructions (i.e. that the user must build MPAS-Ocean with `OPENMP=false` and `USE_LTS=true`) to the user's guide?

gcapodag commented:
@xylar I pushed the latest edits. Please let me know if setting `OPENMP=false` on Chrysalis does not solve the problem. Thanks!

xylar commented:
@gcapodag, without OpenMP I now get NaNs at `2012-10-14_16:39:36`.
```
 ERROR: NaN Detected in state see below for which field contained a NaN.
   -- Statistics information for block fields
     Field: latCell
         Min:   0.111930686000780     
         Max:   0.695048023707948     
     Field: lonCell
         Min:    3.24790359872597     
         Max:    4.97133949663899     
     Field: xCell
         Min:   -6259429.32377882     
         Max:    1264363.43825238     
     Field: yCell
         Min:   -4779483.09361878     
         Max:   -667964.193097126     
     Field: zCell
         Min:    711646.874979247     
         Max:    4080271.51643518     
     Field: areaCell
         Min:    2837243.00641325     
         Max:    14880153386.9070     
 
 
 ERROR: NaN Detected in layerThickness.
   -- Statistics information for layerThickness fields
     Field: seaIcePressure
         Min:   0.000000000000000E+000
         Max:   0.000000000000000E+000
     Field: atmosphericPressure
         Min:   -164.044581752809     
         Max:    634.052276941817     
     Field: rainFlux
         Min:   0.000000000000000E+000
         Max:   0.000000000000000E+000
     Field: evaporationFlux
         Min:   0.000000000000000E+000
         Max:   0.000000000000000E+000
     Field: seaIceFreshWaterFlux
         Min:   0.000000000000000E+000
         Max:   0.000000000000000E+000
 
 
 ERROR: NaN Detected in normalVelocity.
   -- Statistics information for normalVelocity fields
     Field: seaIcePressure
         Min:   0.000000000000000E+000
         Max:   0.000000000000000E+000
     Field: atmosphericPressure
         Min:   -164.044581752809     
         Max:    634.052276941817     
     Field: windStressZonal
         Min:  -0.128912609945117     
         Max:   1.809963222277804E-002
     Field: windStressMeridional
         Min:  -5.610382981044453E-002
         Max:   2.868447792793040E-002
     Field: surfaceStressMagnitude
         Min:   2.133751607774167E-004
         Max:   0.130619092013655     
     Field: surfaceStress
         Min:  -0.118643324612473     
         Max:   0.115021750794859     
     Field: angleEdge
         Min:   -3.13104709207905     
         Max:    3.11419786557324 
```

xylar commented:
I doubt any of that info is helpful but it might at least give you a hint if it's in the LTS region or not.

xylar commented:
What other machines do you have access to?  We don't have Intel support yet on Perlmutter or Chicoma but we do on Compy.

gcapodag commented:
Unfortunately I do not have access to anything else other than Chicoma. I just tried again on Perlmutter with gnu and it worked as expected.

gcapodag commented:
@xylar I found what the problem is. When I was testing yesterday, I had not pulled @mark-petersen 's latest [rebase](https://github.com/E3SM-Project/E3SM/compare/9faa8be181c076ab6eb788384b2364a0d0db52be..1f87513cb8228e47df9bb7d3291d730516c7b595) of the LTS E3SM PR. Unfortunately, that seems to be what is causing the run to die. The namelist for the forward run are different for what concerns the `Rayleigh_drag`, `bottom_drag` and `vegetation_drag`. @xylar can you try and see if the run completes using the LTS E3SM code from this commit (right before the rebase): `9faa8be181c076ab6eb788384b2364a0d0db52be`  ? Thanks.

xylar commented:
@gcapodag, as you presumably noticed in https://github.com/E3SM-Project/E3SM/pull/5460#issuecomment-1548280535, a lot of config options have been renamed in https://github.com/E3SM-Project/E3SM/pull/5595 so you'll need to make the corresponding changes in compass.

I can re-test with the pre-rebase code as soon as I have time.  I'm a bit busy with the Hackathon right now.

gcapodag commented:
> E3SM-Project/E3SM#5595

I was not aware that those changes had been made. Will modify accordingly and retest.

gcapodag commented:
@xylar should be all good now and there is no need for you to try the pre-rebase commit anymore. Thanks!

gcapodag commented:
@xylar @sbrus89 hello folks, is this PR ready to be finalized?

xylar commented:
@gcapodag, I'm hoping to take another look soon.

But this also should ideally wait until https://github.com/E3SM-Project/E3SM/pull/5460 has been merge, and then we update the `E3SM-Project` submodule to include those changes.  Otherwise, it will be hard for anyone but you to know how to run the tests.

xylar reviewed (approved):
Sorry it has taken me awhile to get back to this.  The latest changes look great and I think everything on the compass side is all set to go in.

xylar commented:
@gcapodag, that would be best if he does.  Also, we still need the E3SM changes.  Based on today's MPAS DevOps call, you an @mark-petersen need to chat about whether there's a way to make the number of time levels dynamically allocatable because the alternative (compiling E3SM differently for LTS and for "normal" configurations) is not currently possible and might be a significant effort.

sbrus89 commented:
@gcapodag, really sorry for the delay on this. I'm working on testing now.

sbrus89 reviewed (approved):
Approved after testing on Perlmutter. Note that the analysis step is not working due to https://github.com/E3SM-Project/E3SM/issues/5510.

gcapodag commented:
Thanks @sbrus89 and @xylar !

gcapodag commented:
The LTS E3SM PR has finally been merged so I believe we are ready to merge this one as well, if everyone agrees. @xylar do you need me to do a rebase before we can merge it?

xylar commented:
@gcapodag, given how long it's been since this was created, I think a rebase and a new test using `master` would be a good idea.  We also will need to update the `E3SM-Project` submodule either as part of this PR or in a separate PR.

xylar commented:
I will get the testing started for updating the E3SM-Project submodule.  I think that would be better to do in a separate PR because there have probably been other ocean and/or MPAS framework merges since we last updated.

xylar commented:
E3SM-Project submodule will be updated in https://github.com/MPAS-Dev/compass/pull/651 shortly.

xylar commented:
@gcapodag, Okay, it's updated.  Feel free to rebase, test, post the results of your testing and I'll merge when you've done that.

gcapodag commented:
@xylar , will do, thanks!

gcapodag commented:
Hi @xylar I tested the rebase and everything passed so I pushed it. Please let me know if there is anything left to take care of, thanks.

xylar commented:
Okay, I'll go ahead and merge.  Congratulations, this was a *ton* of work!

gcapodag commented:
Thanks @xylar  and @sbrus89 !
